app/src/main/java/ch/epfl/sdp/peakar/user/profile/NewProfileActivity.java

```diff
@@ -131,4 +192,119 @@ private void expandSelected(boolean expand) {
+
+        // Show edit text
+        EditText changeUsername = findViewById(R.id.profile_username_edit);
+        changeUsername.setVisibility(View.VISIBLE);
+        changeUsername.requestFocus();
+
+        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
+        imm.showSoftInput(changeUsername, InputMethodManager.SHOW_IMPLICIT);
+
+        changeUsername.setOnKeyListener((v, keyCode, event) -> {
+            if ((event.getAction() == KeyEvent.ACTION_DOWN) &&
+                    (keyCode == KeyEvent.KEYCODE_ENTER)) {
+                confirmUsernameButton();
+                return true;
+            }
+            return false;
+        });
+    }
+
+    /**
+     * On confirm username change
+     */
+    public void confirmUsernameButton() {
+        EditText usernameEdit = ((EditText)findViewById(R.id.profile_username_edit));
+        hideKeyboard();
+        String newUsername = usernameEdit.getText().toString();
+        usernameEdit.getText().clear();
+
+        // Start a new thread that will handle the process
+        Thread changeThread = new Thread(){
+            @Override
+            public void run() {
+                // Change the username and wait for the task to end
+                ProfileOutcome result = AuthService.getInstance().getAuthAccount().changeUsername(newUsername);
+
+                // Update the view on the UI thread
+                runOnUiThread(() -> {
+                    ((EditText)findViewById(R.id.profile_username_edit)).getText().clear();
+
+                    // If username has changed, update the username text view
+                    if(result == ProfileOutcome.USERNAME_CHANGED || result == ProfileOutcome.USERNAME_REGISTERED) ((TextView)findViewById(R.id.profile_username)).setText(AuthService.getInstance().getAuthAccount().getUsername());
+                    if(result == ProfileOutcome.USERNAME_REGISTERED) fillListView();
+
+                    // Display the message
+                    Snackbar snackbar = Snackbar.make(findViewById(android.R.id.content), result.getMessage(), Snackbar.LENGTH_LONG);
+                    snackbar.show();
+
+                    // Hide username edit
+                    findViewById(R.id.profile_username_edit).setVisibility(View.GONE);
+
+                    // Show username
+                    findViewById(R.id.profile_username).setVisibility(View.VISIBLE);
+
+                    // Show change button
+                    findViewById(R.id.profile_change_username).setVisibility(View.VISIBLE);
+                });
+            }
+        };
+        changeThread.start();
+    }
+
+    /**
+     * Method to hide the keyboard
+     */
+    private void hideKeyboard() {
+        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
+        imm.hideSoftInputFromWindow(getCurrentFocus().getWindowToken(), 0);
+    }
+
```

Comment:
Add method to detect if keyboard is opened or not.
```suggestion
    /**
     * Checks if the keyboard is opened or not by calculating the difference between the actual
     * height of the profile_collection ListView and the visible height of the View
     *
     * @return true if the keyboard is opened
     */
    private boolean isKeyboardOpened(){
        Rect visibleBounds = new Rect();
        ListView collectionView = findViewById(R.id.profile_collection);
        collectionView.getWindowVisibleDisplayFrame(visibleBounds);
        double heightDiff = collectionView.getHeight() - visibleBounds.height();
        return heightDiff > 0;
    }
```